File: YuXiangROS/Chap5/5.2_5.3_TF_ws/src/demo_python_tf/demo_python_tf/dynamic_tf_broadcaster.py
import rclpy
from geometry_msgs.msg import TransformStamped
from rclpy.node import Node
from tf2_ros import TransformBroadcaster
from tf_transformations import quaternion_from_euler


class DynamicTFBroadcaster(Node):
    def __init__(self):
        super().__init__("dynamic_tf2_broadcaster")

        # 创建动态坐标转换广播器
        self.dynamic_broadcaster_ = TransformBroadcaster(self)
        # 创建定时器，每0.01秒(100Hz)发布一次动态坐标转换消息
        self.timer_ = self.create_timer(0.01, self.publish_dynamic_tf)

    def publish_dynamic_tf(self):
        # 创建动态坐标转换消息
        transform = TransformStamped()

        transform.header.stamp = self.get_clock().now().to_msg()
        transform.header.frame_id = "camera_link"
        transform.child_frame_id = "bottle_link"

        transform.transform.translation.x = 0.2
        transform.transform.translation.y = 0.0
        transform.transform.translation.z = 0.5

        rotation_quat = quaternion_from_euler(0, 0, 0)
        transform.transform.rotation.x = rotation_quat[0]
        transform.transform.rotation.y = rotation_quat[1]
        transform.transform.rotation.z = rotation_quat[2]
        transform.transform.rotation.w = rotation_quat[3]

        # 发布动态坐标转换消息
        self.dynamic_broadcaster_.sendTransform(transform)


def main(args=None):
    rclpy.init(args=args)
    node = DynamicTFBroadcaster()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: YuXiangROS/Chap5/5.2_5.3_TF_ws/src/demo_cpp_tf/src/dynamic_tf_broadcaster.cpp
#include <chrono>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <tf2/LinearMath/Quaternion.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>
#include <tf2_ros/transform_broadcaster.h>

using namespace std::chrono_literals;

class DynamicTFBroadcaster : public rclcpp::Node {
  private:
    // 动态坐标变换广播智能指针
    std::shared_ptr<tf2_ros::TransformBroadcaster> dynamic_broadcaster_;
    // 定时器智能指针
    rclcpp::TimerBase::SharedPtr timer_;

  public:
    DynamicTFBroadcaster() : Node("dynamic_tf_broadcaster") {
        // 创建动态坐标变换广播器
        dynamic_broadcaster_ = std::make_shared<tf2_ros::TransformBroadcaster>(this);
        // 创建定时器, 10ms 周期(100Hz)
        timer_ = this->create_wall_timer(
            10ms,
            std::bind(&DynamicTFBroadcaster::publish_dynamic_tf_, this)
        );
    }

  private:
    void publish_dynamic_tf_() {
        geometry_msgs::msg::TransformStamped transform;

        transform.header.stamp = this->get_clock()->now();
        transform.header.frame_id = "map";
        transform.child_frame_id = "base_link";

        transform.transform.translation.x = 2.0;
        transform.transform.translation.y = 3.0;
        transform.transform.translation.z = 0.0;

        tf2::Quaternion quat;
        quat.setRPY(0.0, 0.0, 30 * M_PI / 180);
        transform.transform.rotation = tf2::toMsg(quat);

        dynamic_broadcaster_->sendTransform(transform);
    }
};

int main(int argc, char* argv[]) {
    rclcpp::init(argc, argv);
    auto node = std::make_shared<DynamicTFBroadcaster>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}